app/src/org/commcare/dalvik/activities/HomeActivityUIController.java

```diff
@@ -0,0 +1,404 @@
+
+    private void refreshVersionText() {
+        TextView version = (TextView)activity.findViewById(R.id.str_version);
+        if (version == null) {
+            return;
+        }
+        version.setText(CommCareApplication._().getCurrentVersionString());
+    }
+
+    private void refreshButtonTextSources() {
+        if (activity.isDemoUser()) {
+            syncKey = "home.sync.demo";
+            lastMessageKey = "home.sync.message.last";
+            homeMessageKey = "home.start.demo";
+            logoutMessageKey = "home.logout.demo";
+        } else {
+            syncKey = "home.sync";
+            lastMessageKey = "home.sync.message.last";
+            homeMessageKey = "home.start";
+            logoutMessageKey = "home.logout";
+        }
+    }
+
+    private void refreshHomeAndLogoutButtons() {
+        if (startButton != null) {
+            startButton.setText(Localization.get(homeMessageKey));
+        }
+        if (logoutButton != null) {
+            logoutButton.setText(Localization.get(logoutMessageKey));
+        }
+    }
+
+    /**
+     * Call this method before a new isolated instance of using any of the 3 fields for which it
+     * obtains values
+     */
+    private void refreshDataFromSyncDetails() {
+        try {
+            Pair<Long, int[]> syncDetails = CommCareApplication._().getSyncDisplayParameters();
+            this.lastSyncTime = syncDetails.first;
+            this.numUnsentForms = syncDetails.second[0];
+            this.numIncompleteForms = syncDetails.second[1];
+        } catch (UserStorageClosedException e) {
+            activity.returnToLogin();
+            return;
+        }
+    }
+
+    private void refreshSyncButton() {
+        if (syncButton != null) {
+            setSyncButtonText(syncKey);
+        }
+    }
+
+    private void showSyncMessage() {
+        CharSequence syncTimeMessage;
+        if (lastSyncTime == 0) {
+            syncTimeMessage = Localization.get("home.sync.message.last.never");
+        } else {
+            syncTimeMessage = DateUtils.formatSameDayTime(lastSyncTime, new Date().getTime(), DateFormat.DEFAULT, DateFormat.DEFAULT);
+        }
+
+        String message = "";
+        if (numUnsentForms == 1) {
+            message += Localization.get("home.sync.message.unsent.singular") + "\n";
+        } else if (numUnsentForms > 1) {
+            message += Localization.get("home.sync.message.unsent.plural", new String[]{String.valueOf(numUnsentForms)}) + "\n";
+        }
+
+        message += Localization.get(lastMessageKey, new String[]{syncTimeMessage.toString()});
+        displayMessage(message, syncNotOk(), true);
+    }
+
+    protected void displayMessage(String message, boolean bad, boolean suppressToast) {
+        if (!suppressToast) {
+            Toast.makeText(activity, message, Toast.LENGTH_LONG).show();
+        }
+        adapter.setNotificationTextForButton(R.layout.home_sync_button, message);
+    }
+
+    private boolean syncNotOk() {
```

Comment:
I think this name could be a bit less concerning. It's not that the sync state isn't ok, it's just that the user should be notified that it's been a while.